TLS v1.3: Support a stateful ticket and test HAVE_EXT_CACHE

tmael opened the pull request:
# Description

This PR adds support for TLSv1.3 stateful session tickets when using SSL_OP_NO_TICKET and corrects a couple of issues:

1) The callback set using `wolfSSL_CTX_sess_set_remove_cb` will only be called when a session is being evicted from cache or when the session times out
2) Cache miss
3) SessionCache leak

`AddSessionToCache() `was removing a session from the external cache. 
The internal `SessionCache` would get updated but the external cache wasn't being called again.

Additionally, when overwriting a session, AddSessionToCache() was incrementing the `totalCount` and the `nextIndx` where to place the next session item leading to a `SessionCache` leak


Fixes zd#15434

# Testing

With a new `test_wolfSSL_CTX_add_session_ext` to test the external session cache for TLSv1.1, v1.2, and v1.3 with various cache configurations.

```
./configure --enable-debug --disable-shared -enable-haproxy --enable-quic --enable-sp=yes,asm --enable-intelasm --enable-sp-math CFLAGS="-DENABLE_SESSION_CACHE_ROW_LOCK" && make check

cd $HOME/wolfssl
sed -E --in-place 's/wolfSSL_CTX_no_ticket_TLSv13\(ctx\);$/wolfSSL_CTX_set_options\(ctx, SSL_OP_NO_TICKET\);/' examples/server/server.c || exit $?
./configure --enable-session-ticket --enable-opensslextra && make
./examples/server/server -v 4 -T -i
./examples/client/client -v 4 
AND
openssl s_client -connect localhost:11111 -tls1_3 -cert ./certs/client-cert.pem -key ./certs/client-key.pem -CAfile ./certs/ca-cert.pem  < /dev/null
```
# Checklist

 - [x] added tests
 - [ ] updated/added doxygen
 - [ ] updated appropriate READMEs
 - [ ] Updated manual and documentation

tmael commented on src/ssl.c:
```diff
@@ -14721,7 +14721,6 @@ void AddSession(WOLFSSL* ssl)
         idSz = session->sessionIDSz;
     }
 
-    session->timeout = ssl->timeout;
```
Removed redundant line since there's another `session->timeout = ssl->timeout;` ~13 lines below.

SparkiDev commented:
How did you test?
Has it been tried against OpenSSL?
What about the DoTls13SessionTicket?

tmael commented:
> How did you test? Has it been tried against OpenSSL? What about the DoTls13SessionTicket?

@SparkiDev 
Tested with the following.  From OpenSSL client, checked the length of the session ticket is always 32.
DoTls13SessionTicket would fail and a new handshake is done.
 ```
cd $HOME/wolfssl
sed -E --in-place 's/wolfSSL_CTX_no_ticket_TLSv13\(ctx\);$/wolfSSL_CTX_set_options\(ctx, SSL_OP_NO_TICKET\);/' examples/server/server.c || exit $?
./configure --enable-session-ticket --enable-opensslextra && make
./examples/server/server -v 4 -T -i
./examples/client/client -v 4 
AND
openssl s_server -key ./certs/server-key.pem -cert ./certs/server-cert.pem -accept 11111 -www
```

julek-wolfssl commented on src/tls13.c:
```diff
@@ -9690,12 +9690,22 @@ static int SendTls13NewSessionTicket(WOLFSSL* ssl)
 #else
     extSz = EXTS_SZ;
 #endif
+    if (IsAtLeastTLSv1_3(ssl->version) &&
```
By definition this function only gets called for TLS 1.3. No need to check for this.

julek-wolfssl commented on src/tls13.c:
```diff
@@ -9720,11 +9730,31 @@ static int SendTls13NewSessionTicket(WOLFSSL* ssl)
     /* ticket */
-    XMEMCPY(output + idx, ssl->session->ticket, ssl->session->ticketLen);
-    idx += ssl->session->ticketLen;
+    if (IsAtLeastTLSv1_3(ssl->version) &&
```
Same TLS 1.3 comment as above.

julek-wolfssl commented on src/tls13.c:
```diff
@@ -9720,11 +9730,31 @@ static int SendTls13NewSessionTicket(WOLFSSL* ssl)
+        if (ssl->session->haveAltSessionID == 1)
+            XMEMCPY(output + idx, ssl->session->altSessionID, ID_LEN);
+        else
+            XMEMCPY(output + idx, ssl->session->sessionID, ID_LEN);
```
We can NOT use the session ID for the ID. It is user provided and can be anything (for example a string of `0`'s).

julek-wolfssl reviewed (changes requested):
I don't think checking `ssl->options.mask & WOLFSSL_OP_NO_TICKET` inside of `SendTls13NewSessionTicket` makes sense. It would be better to check this in locations where this function is called like in `wolfSSL_accept_TLSv13`.

julek-wolfssl commented:
Maybe also only call the `AddSession` in `SendTls13NewSessionTicket` when `ssl->session->haveAltSessionID` is true?

julek-wolfssl commented:
`DoClientTicket` needs to be updated to understand the 32 byte ticket.

tmael commented:
@SparkiDev, @julek-wolfssl  Addressed review feedback. 
The updated solution would allow a session resumption.
With the first solution, a server sends a `sessionID` (copy of `altSessionID`) and does a full handshake, essentially ignoring the session ID from the client.

julek-wolfssl commented on tests/api.c:
```diff
@@ -7645,6 +7645,538 @@ static int test_wolfSSL_CTX_add_session(void)
+
+static hashTable server_sessionCache;
+
+static int new_sessionCb(WOLFSSL *ssl, WOLFSSL_SESSION *sess)
```
Please namespace these functions a little with a common unique prefix. Unfortunately C doesn't provide us a way to namespace and this is global to the entire `api.c`.

tmael replied:
@julek-wolfssl, Good suggestion. I initially wanted to avoid cluttered code with long function names for clarity but I could use your short prefix.

julek-wolfssl reviewed (changes requested):
Very nice! I especially like the testing. I have some comments though.

tmael commented on src/ssl.c:
```diff
@@ -15212,6 +15217,17 @@ int AddSessionToCache(WOLFSSL_CTX* ctx, WOLFSSL_SESSION* addSession,
+                else {
+                    ret = 0; /* reset to success */
+                }
+            }
```
@julek-wolfssl, what do we need to do here?

julek-wolfssl replied:
Don't do any `new` or `free` calls when holding the `SESSION_ROW_WR_LOCK(sessRow)` lock. If you need to do this, save the pointer and perform the `free` call after you unlock the lock.

julek-wolfssl commented on src/ssl.c:
```diff
@@ -15197,10 +15200,12 @@ int AddSessionToCache(WOLFSSL_CTX* ctx, WOLFSSL_SESSION* addSession,
+            if (sessRow->totalCount < SESSIONS_PER_ROW)
+                sessRow->totalCount++;
+            sessRow->nextIdx = (sessRow->nextIdx + 1) % SESSIONS_PER_ROW;
+        }
```
Nevermind this. We don't use `sessRow->nextIdx` when `overwrite`. This change is correct. Thank @tmael .

julek-wolfssl reviewed (changes requested):
Added comments.

dgarske commented on tests/api.c:
```diff
@@ -7447,6 +7447,592 @@ static int test_wolfSSL_CTX_add_session(void)
+static int twcase_new_sessionCb(WOLFSSL *ssl, WOLFSSL_SESSION *sess)
+{
+    (void)ssl;
+    int i;
```
```tests/api.c: In function ‘twcase_new_sessionCb’:
tests/api.c:7488:5: error: ISO C90 forbids mixed declarations and code [-Werror=declaration-after-statement]
 7488 |     int i;
      |     ^~~
```

dgarske commented on tests/api.c:
```diff
@@ -7447,6 +7447,592 @@ static int test_wolfSSL_CTX_add_session(void)
+{
+    (void)ctx;
+    (void)sess;
+    int i;
```
```tests/api.c: In function ‘twcase_remove_sessionCb’:
tests/api.c:7531:5: error: ISO C90 forbids mixed declarations and code [-Werror=declaration-after-statement]
 7531 |     int i;
      |     ^~~
```

dgarske commented on tests/api.c:
```diff
@@ -7447,6 +7447,592 @@ static int test_wolfSSL_CTX_add_session(void)
+    (void)ssl;
+    (void)id;
+    (void)len;
+    int i;
```
```tests/api.c: In function ‘twcase_get_sessionCb’:
tests/api.c:7568:5: error: ISO C90 forbids mixed declarations and code [-Werror=declaration-after-statement]
 7568 |     int i;
      |     ^~~```

dgarske commented on tests/api.c:
```diff
@@ -7447,6 +7447,592 @@ static int test_wolfSSL_CTX_add_session(void)
+    if (wc_LockMutex(&server_sessionCache.htLock) != 0) {
+        return;
+    }
+    for(i = 0; i < SESSION_CACHE_SIZE; i++) {
```
NIT: `if ` and `for (`

dgarske commented:
@SparkiDev please review also. Thanks

julek-wolfssl commented on src/dtls.c:
```diff
@@ -370,22 +370,25 @@ static int TlsSessionIdIsValid(const WOLFSSL* ssl, WolfSSL_ConstVector sessionID
-                if (IsAtLeastTLSv1_3(sess->version))
-                    wolfSSL_FreeSession(ssl->ctx, sess);
+                if (IsAtLeastTLSv1_3(extSess->version)) {
+                    if (!copy)
```
Actually this section is invalid. `TlsSessionIdIsValid` should never handle TLS 1.3 code. The legacy session code should never match a TLS 1.3 session. If it does, we should treat this as if we didn't find the session.

tmael replied:
I am not completely certain that I understood your comment here. Attempted to fix the issue.

julek-wolfssl replied:
Fix is good.

julek-wolfssl commented on src/ssl.c:
```diff
@@ -15479,34 +15593,39 @@ int AddSessionToCache(WOLFSSL_CTX* ctx, WOLFSSL_SESSION* addSession,
+        cacheSession = (WOLFSSL_SESSION*) XMALLOC(sizeof(WOLFSSL_SESSION),
+                                         sessRow->heap, DYNAMIC_TYPE_SESSION);
+        if (cacheSession == NULL) {
+            return MEMORY_E;
```
And not releasing the session lock. Good catch.

julek-wolfssl commented on src/ssl.c:
```diff
@@ -15479,34 +15593,39 @@ int AddSessionToCache(WOLFSSL_CTX* ctx, WOLFSSL_SESSION* addSession,
+        /* Figure out who owns the ex_data */
+        if (cacheSession->ownExData) {
+            /* Prioritize cacheSession copy */
+            XMEMCPY(&addSession->ex_data, &cacheSession->ex_data,
```
`overwrite == 1` means that the session already exists in cache. The `ex_data` in cache is the single source of the real `ex_data`. It can be shared between objects.

SparkiDev replied:
I understand. Thanks.

SparkiDev commented on src/ssl.c:
```diff
@@ -15086,24 +15197,23 @@ int wolfSSL_SetSession(WOLFSSL* ssl, WOLFSSL_SESSION* session)
         }
     }
 
-    if (ret == WOLFSSL_SUCCESS && SslSessionCacheOff(ssl, session)) {
```
I understand. Thanks.

tmael commented:
retest this please

tmael commented:
retest this please

julek-wolfssl reviewed (changes requested):
I believe that this PR is introducing an async failure in DTLS 1.3. Please investigate with debug logs to attempt to narrow down the issue.
